src/Builder/BuilderEncoder.php

```diff
@@ -0,0 +1,293 @@
+class BuilderEncoder implements Encoder
+{
+    use EncodeIfSupported;
+
+    /**
+     * {@inheritdoc}
+     */
+    public function canEncode($value): bool
+    {
+        return $value instanceof Pipeline
+            || $value instanceof OperatorInterface
+            || $value instanceof ExpressionInterface
+            || $value instanceof QueryInterface
+            || $value instanceof FieldQueryInterface
+            || $value instanceof AccumulatorInterface
+            || $value instanceof ProjectionInterface
+            || $value instanceof WindowInterface;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function encode($value): stdClass|array|string
+    {
+        if (! $this->canEncode($value)) {
+            throw UnsupportedValueException::invalidEncodableValue($value);
+        }
+
+        // A pipeline is encoded as a list of stages
+        if ($value instanceof Pipeline) {
+            $encoded = [];
+            foreach ($value->getIterator() as $stage) {
+                $encoded[] = $this->encodeIfSupported($stage);
+            }
+
+            return $encoded;
+        }
+
+        // This specific encoding code if temporary until we have a generic way to encode stages and operators
+        if ($value instanceof FieldPathInterface) {
+            return '$' . $value->name;
+        }
+
+        if ($value instanceof Variable) {
+            return '$$' . $value->name;
+        }
+
+        if ($value instanceof QueryObject) {
+            return $this->encodeQueryObject($value);
+        }
+
+        if ($value instanceof CombinedFieldQuery) {
+            return $this->encodeCombinedFilter($value);
+        }
+
+        if ($value instanceof OutputWindow) {
+            return $this->encodeOutputWindow($value);
+        }
+
+        if (! $value instanceof OperatorInterface) {
+            throw new LogicException(sprintf('Class "%s" does not implement OperatorInterface.', $value::class));
+        }
+
+        // The generic but incomplete encoding code
+        switch ($value::ENCODE) {
+            case Encode::Single:
+                return $this->encodeAsSingle($value);
+
+            case Encode::Array:
+                return $this->encodeAsArray($value);
+
+            case Encode::Object:
+                return $this->encodeAsObject($value);
+
+            case Encode::Group:
+                assert($value instanceof GroupStage);
+
+                return $this->encodeAsGroup($value);
+        }
+
+        throw new LogicException(sprintf('Class "%s" does not have a valid ENCODE constant.', $value::class));
+    }
+
+    /**
+     * Encode the value as an array of properties, in the order they are defined in the class.
+     */
+    private function encodeAsArray(OperatorInterface $value): stdClass
+    {
+        $result = [];
+        /** @var mixed $val */
+        foreach (get_object_vars($value) as $val) {
```

Comment:
Is this fragile in that it depends on the declaration order of class properties? I suppose if the classes are all generated from YAML this won't be a concern.

Reply:
Yes, we use the order in which properties are declared in the object to generate the array. Otherwise, we have to generate a specific encoding class for each operator with all the necessary metadata.